openmc/source.py

```diff
@@ -333,6 +380,158 @@ def from_xml_element(cls, elem: ET.Element, meshes=None) -> 'openmc.Source':
         return source
 
 
+class CompiledSource(Source):
+    """
+    A class used to apply a compiled source library
+    """
+    def __init__(self, library: Optional[str] = None, strength=1.0) -> None:
+        super().__init__(strength=strength)
+
+        self._library = None
+        if library is not None:
+            self.library = library
+
+        self._parameters = None
+
+    @property
+    def type(self) -> str:
+        return "compiled"
+
+    @property
+    def library(self) -> str:
+        return self._library
+
+    @property
+    def parameters(self) -> str:
+        return self._parameters
+
+    @library.setter
+    def library(self, library_name):
+        cv.check_type('library', library_name, str)
+        self._library = library_name
+
+    @parameters.setter
+    def parameters(self, parameters_path):
+        cv.check_type('parameters', parameters_path, str)
+        self._parameters = parameters_path
+
+    def populate_xml_element(self, element):
+        """Add necessary compiled source information to an XML element
+
+        Returns
+        -------
+        element : lxml.etree._Element
+            XML element containing source data
+
+        """
+        super().populate_xml_element(element)
+
+        element.set("library", self.library)
+
+        if self.parameters is not None:
+            element.set("parameters", self.parameters)
+
+    @classmethod
+    def from_xml_element(cls, elem: ET.Element, meshes=None) -> 'openmc.CompiledSource':
+        """Generate a compiled source from an XML element
+
+        Parameters
+        ----------
+        elem : lxml.etree._Element
+            XML element
+        meshes : dict
+            Dictionary with mesh IDs as keys and openmc.MeshBase instaces as
+            values
+
+        Returns
+        -------
+        openmc.Source
+            Source generated from XML element
+
+        """
+        library = get_text(elem, 'library')
+
+        source = cls(library)
+
+        strength = get_text(elem, 'strength')
+        if strength is not None:
+            source.strength = float(strength)
+
+        parameters = get_text(elem, 'parameters')
+        if parameters is not None:
+            source.parameters = parameters
+
+        return source
+
+
+class FileSource(Source):
+    """
+    """
+    def __init__(self, filename: Optional[str] = None, strength=1.0) -> None:
+        super().__init__(strength=strength)
+
+        self._file = None
+
+        if filename is not None:
+            self.file = filename
```

Comment:
Maybe I'm missing something subtle about how this all works, but shouldn't this be:
```suggestion
            self._file = filename
```